test/Dialect/FIRRTL/lower-layers.mlir

```diff
@@ -1,11 +1,366 @@
+  // CHECK:   firrtl.ref.define %[[p]], %1 : !firrtl.probe<uint<1>>
+  // CHECK: }
+  // CHECK: firrtl.module @NestedCaptureHardware() {
+  // CHECK:   %[[b1:.+]], %[[b2:.+]] = firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A_B.sv", excludeFromFileList>} @[[B]]
+  // CHECK:   %[[a1:.+]], %[[a2:.+]] = firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A.sv", excludeFromFileList>} @[[A]]
+  // CHECK:   %0 = firrtl.ref.resolve %[[a2]] : !firrtl.probe<uint<1>>
+  // CHECK:   firrtl.strictconnect %[[b1]], %0 : !firrtl.uint<1>
+  // CHECK:   %1 = firrtl.ref.resolve %[[a1]] : !firrtl.probe<uint<1>>
+  // CHECK:   firrtl.strictconnect %[[b2]], %1 : !firrtl.uint<1>
+  // CHECK: }
+  firrtl.module @NestedCaptureHardware() {
+    firrtl.layerblock @A {
+      %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+      %c1_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+      firrtl.layerblock @A::@B {
+        %0 = firrtl.add %c0_ui1, %c1_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+      }
+    }
+  }
+
+  // CHECK: firrtl.module private @[[A:.+]](in %[[p:.+]]: !firrtl.uint<1>) {
+  // CHECK:   %c1_ui1 = firrtl.constant 1 : !firrtl.uint<1>
+  // CHECK:   firrtl.when %[[p]] : !firrtl.uint<1> {
+  // CHECK:     %0 = firrtl.add %[[p]], %c1_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+  // CHECK:   }
+  // CHECK: }
+  // CHECK: firrtl.module @WhenUnderLayer() {
+  // CHECK:   %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %[[p:.+]] = firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A.sv", excludeFromFileList>} @[[A]]
+  // CHECK:   firrtl.strictconnect %[[p]], %c0_ui1 : !firrtl.uint<1>
+  // CHECK: }
+  firrtl.module @WhenUnderLayer() {
+    %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      %c1_ui1 = firrtl.constant 1 : !firrtl.uint<1>
+      firrtl.when %c0_ui1 : !firrtl.uint<1> {
+        %0 = firrtl.add %c0_ui1, %c1_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+      }
+    }
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Connecting/Defining Refs
+  //===--------------------------------------------------------------------===//
+
+  // Src and Dst Outside Layerblock.
+  //
+  // CHECK: firrtl.module private @[[A:.+]]() {
+  // CHECK: }
+  // CHECK: firrtl.module @SrcDstOutside() {
+  // CHECK:   %0 = firrtl.wire : !firrtl.probe<uint<1>>
+  // CHECK:   %1 = firrtl.wire : !firrtl.probe<uint<1>>
+  // CHECK:   firrtl.ref.define %1, %0 : !firrtl.probe<uint<1>>
+  // CHECK:   firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A.sv", excludeFromFileList>} @[[A:.+]]()
+  // CHECK: }
+  firrtl.module @SrcDstOutside() {
+    %0 = firrtl.wire : !firrtl.probe<uint<1>, @A>
+    %1 = firrtl.wire : !firrtl.probe<uint<1>, @A>
+    firrtl.layerblock @A {
+      firrtl.ref.define %1, %0 : !firrtl.probe<uint<1>, @A>
+    }
+  }
```

Comment:
This is interesting as it's actually fine to do. However, there is a question of how to avoid unnecessarily creating the module.